oop-object_interactions-medium-q2 (#683)
canvas-gamification/canvas-gamification-junit-tests: src/oop/object_interactions/medium/q2/QuestionText.md

# oop-object_interactions-medium-q2

- Java question

## Question Text

> Many people keep secrets that they only share with a trusted friend. Make a class called **Person** that has a secret 

Reviewer: Please add that they also need a second class; otherwise they might miss it at the bottom.

# oop-object_interactions-medium-q2

- Java question

## Question Text

Many people keep secrets that they only share with a trusted friend. Make a class called **Person** that has a secret

> and potentially knows about the secrets of his/her friends. This class has several attributes: **myName**, **mySecret**,

Reviewer 2: ```suggestion
and potentially knows about the secrets of their friends. This class has several attributes: **myName**, **mySecret**,
```

a list of names of **friends**, and a list of **secrets** from those friends.

## Person Class

The `Person` class currently has the following items:

- Attributes:
    - A `String` called **myName**
    - A `String` called **mySecret**
    - An array of `String` called **friends**

>     - An array of `String` called **secrets**

Reviewer 2: these two variable names are inconsistent in this file. Here they are "friends" and "secrets" but farther down it is "friendNames" and "friendSecrets"

    - None of these attributes should be accessible from other classes.

- Constructors:
    - a constructor that takes a `String` for initializing the **myName** attribute, another `String` for initializing
      the **mySecret** attribute, and an array of `String` for initializing the **friendNames** attribute.
    - It should initialize the **friendSecrets** array as an array of null `String` with the size of **friendNames**
      attribute that is previously initialized with the input array of `String`.
    - It should be accessible from other classes.

- Methods:
    - Getter methods for **myName** and **mySecret** attributes that are accessible from other classes.

You must add the following items to the class:

- Methods:

>     - A `hearSecret` method that takes two `String`. One is the name of a friend and the other is that friend's secret.

Reviewer: ```suggestion
    - A `hearSecret` method that has two `String` parameters. The first one is the name of a friend and the second is the friend's secret.
```

>     - A `hearSecret` method that take two `String` one is the name of a friend and the other is that friends secret.

Reviewer 2: ```suggestion
    - A `hearSecret` method that takes two `String`. One is the name of a friend and the other is that friend's secret.
```

>       Your method should find the friend using the names in the **friendNames** attribute and it should change the

Reviewer: ```suggestion
      Your method should try to find the friend using the names in the **friendNames** attribute and it should change the
```

>       Your methods should find the friends using the input name in the **friendNames** attribute and if the index of

Reviewer 2: ```suggestion
      Your method should find the friend using the input name in the **friendNames** attribute and if the index of
```

>       value of **friendSecrets** at the same index as the friend is found in **friendNames**, if you do not find the

Reviewer: ```suggestion
      value of **friendSecrets** at the same index as the friend is found in **friendNames**. If you do not find the
```

>       friend is i, it should change the value of **friendSecrets** at i to the value of the input for the secret.

Reviewer 2: this wording is a bit awkward. Maybe change to something like, "**friendNames** attribute and it should change the value of **friendSecrets** at the same index as the friend is located to the value of the input for the secret."

>       name you must not change anythim.

Reviewer: ```suggestion
      name you must not change anything.
```

>     - A `writeInDiary` method which prints `I have the following secrets:` and then print each secret that the

Reviewer 2: ```suggestion
    - A `writeInDiary` method which prints `I have the following secrets:` and then prints each secret that the
```

>       **Person** knows in each line. It should write the secret in the following format:

Reviewer 2: ```suggestion
      **Person** knows on a separate line. Given a friend `Ginger` and a secret `she is scared of bears` it should write the secret in the following format:
```

Reviewer 2: then get rid of the below line since it got merged above

>     - A `writeInDiary` method which prints `I have the following secrets:` in first line and then prints each secret
>       that the **Person** knows on a separate line only if there is a secret. Given a friend `Ginger` and a secret
>       `she is scared of bears` it should write the secret in the following format: `Ginger told me she is scared of bears.`

Reviewer: ```suggestion
    - A `writeInDiary` method which prints `I have the following secrets:` in first line and then prints each secret
      that the **Person** knows on a separate line. For each secret they know, you must print the name of the friend and what you were told the secret is, with the format "<**Friend**> told me <**secret**>.". For example, given a friend `Ginger` and a secret `she is scared of bears` it should print the secret in the following format: `Ginger told me she is scared of bears.`
```

## TestFriendship Class

Your `TestFriendship` class must have the following items:

- Main Method:
    - A `main` method which initializes two **Person** objects, one with a **myName** value of `Ann` and **mySecret**
      value of `she hates chocolates` and the other with values of `Cam` and `she is scared of heights`.

>     - Ann is friends with Bob and Cam in that order and Cam is only friends with Ann.

Reviewer 2: I feel like since this is part of the constructor information it should be provided int the above point

      the **mySecret** attribute, and an array of `String` for initializing the **friendNames** attribute.
    - It should initialize the **friendSecrets** array as an array of null `String` with the size of **friendNames**
      attribute that is previously initialized with the input array of `String`.
    - It should be accessible from other classes.

- Methods:
    - Getter methods for **myName** and **mySecret** attributes that are accessible from other classes.

You must add the following items to the class:

- Methods:
    - A `hearSecret` method that takes two `String`. One is the name of a friend and the other is that friend's secret.
      Your method should find the friend using the input name in the **friendNames** attribute and it should change the
      value of **friendSecrets** at the same index as the friend is located to the value of the input for the secret.
    - A `writeInDiary` method which prints `I have the following secrets:` and then prints each secret that the
      **Person** knows on a separate line. Given a friend `Ginger` and a secret `she is scared of bears` it should write
      the secret in the following format:

>       if friend `Ginger` has the secret `she is scared of bears` it should be

Reviewer 2: ```suggestion
```

>       like `Ginger told me she is scared of bears.`

Reviewer 2: ```suggestion
      `Ginger told me she is scared of bears.`
```
